# Contents
1. Importing libraries and dataframes.
2. Creating a dataframe.
3. Creating a mix typed column.
4. Checking for mixed types.
5. Finding missing values.
6. Addressing missing values.
7. Finding duplicates.
8. Task 5
    1. Question 2
    2. Question 3
    3. Question 4
    4. Question 5
    5. Question 6
    6. Question 7 and 8
    7. Question 9

# 01. Importing libraries

In [2]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [3]:
#Importing dataframes
path = r'C:\Users\Owner\OneDrive\Desktop\Data Analytics\Achievement 4\11-2022 Instacart Basket Analysis'

In [4]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [7]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [8]:
#Create a dataframe
df_test = pd.DataFrame()

In [9]:
#Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [10]:
#Check for mixed types
for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc
            [0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print (col)

mix


In [11]:
df_test['mix'] = df_test['mix'].astype('str')

In [12]:
#Finding missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [13]:
df_nan = df_prods[df_prods['product_name'].isnull() ==  True]

In [14]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [17]:
#Addressing missing values
df_prods.shape

(49693, 5)

In [20]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() ==  False]

In [21]:
df_prods_clean.shape


(49677, 5)

In [22]:
#Finding duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [23]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [24]:
df_prods_clean.shape

(49677, 5)

In [25]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [26]:
df_prods_clean_no_dups.shape

(49672, 5)

In [27]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

# Achievement 4 Task 5

# Question 2

In [28]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

Prices should be investigated further as the maximum value is 99999. That cannot be corrected. This would throw off the mean value, too.

# Question 3 and 4

In [30]:
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc
            [0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print (col)

#There are no mixed data type columns in df_ords

# Question 5

In [32]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
customer_id                    0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

The only missing values are in 'days_since_prior_order', and I would assume these entries are missing because it was the customers' first orders with Instacart.

# Question 6

In [33]:
df_ords.shape

(3421083, 8)

In [34]:
df_ords_clean = df_ords[df_ords['days_since_prior_order'].isnull() ==  False]

In [35]:
df_ords_clean.shape

(3214874, 8)

I chose to create a new data frame to remove the missing values so as not to disturb the database. We would eventually need these values if a customer decided to use Instacart again. This value would not be missing then.

# Question 7 and 8

In [36]:
df_ordsdups = df_ords_clean[df_ords_clean.duplicated()]

In [37]:
df_ordsdups

Empty DataFrame
Columns: [Unnamed: 0, order_id, customer_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [38]:
df_ords_clean.shape

(3214874, 8)

In [39]:
df_ords_clean_no_dups = df_ords_clean.drop_duplicates()

In [42]:
df_ords_clean_no_dups.shape

(3214874, 8)

There are no duplicates in df_ords.

# Question 9

In [43]:
df_ords_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))